Exploring DataFrame information
Learning goal: By the end of this subsection, you should be comfortable finding general information about the data stored in pandas DataFrames.

Once you have loaded your data into pandas, it will more likely than not be in a DataFrame. However, if the data set in your DataFrame has 60,000 rows and 400 columns, how do you even begin to get a sense of what you're working with? Fortunately, pandas provides some convenient tools to quickly look at overall information about a DataFrame in addition to the first few and last few rows.
複雑すぎるデータだとどこから手を付ければいいかわからない、pandasの便利な道具をつかえば理解しやすくなる。

In order to explore this functionality, we will import the Python scikit-learn library and use an iconic dataset that every data scientist has seen hundreds of times: British biologist Ronald Fisher's Iris data set used in his 1936 paper "The use of multiple measurements in taxonomic problems":
scikit-learnを使う、アヤメデータセットはデータサイエンティストなら何回も目にするだろう、今回はBritish biologist Ronald Fisher's Iris dataをつかうよ
more likely than おそらく

In [2]:
import sys
!{sys.executable} -m pip install scikit-learn pandas


[notice] A new release of pip is available: 24.0 -> 26.1.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import pandas as pd
from sklearn.datasets import load_iris#ただのデータセット
iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])
iris_df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2
...,...,...,...,...
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3


DataFrame.shape
We have loaded the Iris Dataset in the variable iris_df. Before diving into the data, it would be valuable to know the number of datapoints we have and the overall size of the dataset. It is useful to look at the volume of data we are dealing with.
データを読み込んだから、全体の大きさを知るろう
diving into飛び込む

In [4]:
iris_df.shape#shapeは教えてくれるだけで編集はしないのでパンダズのアトリビュート、メソッドではない

(150, 4)

So, we are dealing with 150 rows and 4 columns of data. Each row represents one datapoint and each column represents a single feature associated with the data frame. So basically, there are 150 datapoints containing 4 features each.
150のデータポイントに4つの特徴があるデータ
shape here is an attribute of the dataframe and not a function, which is why it doesn't end in a pair of parentheses.
shapeは関数じゃないよ
attribute情報
parentheses丸カッコ
end inで終わる
, which is whyだから～なのです。

DataFrame.columns
Let us now move into the 4 columns of data. What does each of them exactly represent? The columns attribute will give us the name of the columns in the dataframe.

In [5]:
iris_df.columns#カラムの名前を参照してくれる

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='str')

As we can see, there are four(4) columns. The columns attribute tells us the name of the columns and basically nothing else. This attribute assumes importance when we want to identify the features a dataset contains.
nothing elseそれ以外何もない

DataFrame.info
The amount of data(given by the shape attribute) and the name of the features or columns(given by the columns attribute) tell us something about the dataset. Now, we would want to dive deeper into the dataset. The DataFrame.info() function is quite useful for this.

In [6]:
iris_df.info()#今までのアトリビュートと「データの種類」や「穴あき（欠損値）がないか」までを1枚のレポートにまとめて教えてくれる機能

#Non-Nullはネルが一個もない

<class 'pandas.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


From here, we get to can make a few observations:

The DataType of each column: In this dataset, all of the data is stored as 64-bit floating-point numbers.
Number of Non-Null values: Dealing with null values is an important step in data preparation. It will be dealt with later in the notebook.
DataFrame.describe()
Say we have a lot of numerical data in our dataset. Univariate statistical calculations such as the mean, median, quartiles etc. can be done on each of the columns individually. The DataFrame.describe() function provides us with a statistical summary of the numerical columns of a dataset.
mean平均
numerical数値の
Univariate一つの変数
Sayたとえば〜だとしましょう
can be done行うことができる
on対して
a statistical summary統計的な要約
laterあとで


In [ ]:
iris_df.describe()#平均、中央値、四分位数、合計をすべて計算して表にしてくれる

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
count,150.000000,150.000000,150.000000,150.000000
mean,5.843333,3.057333,3.758000,1.199333
std,0.828066,0.435866,1.765298,0.762238
min,4.300000,2.000000,1.000000,0.100000
25%,5.100000,2.800000,1.600000,0.300000
50%,5.800000,3.000000,4.350000,1.300000
75%,6.400000,3.300000,5.100000,1.800000
max,7.900000,4.400000,6.900000,2.500000


The output above shows the total number of data points, mean, standard deviation, minimum, lower quartile(25%), median(50%), upper quartile(75%) and the maximum value of each column.

DataFrame.head
With all the above functions and attributes, we have got a top level view of the dataset. We know how many data points are there, how many features are there, the data type of each feature and the number of non-null values for each feature.

Now its time to look at the data itself. Let's see what the first few rows(the first few datapoints) of our DataFrame look like:

In [8]:
iris_df.head()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


As the output here, we can see five(5) entries of the dataset. If we look at the index at the left, we find out that these are the first five rows.

Exercise:
From the example given above, it is clear that, by default, DataFrame.head returns the first five rows of a DataFrame. In the code cell below, can you figure out a way to display more than five rows?

DataFrame.tail
Another way of looking at the data can be from the end(instead of the beginning). The flipside of DataFrame.head is DataFrame.tail, which returns the last five rows of a DataFrame:

In [9]:
iris_df.tail()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3
149,5.9,3.0,5.1,1.8


Missing Data
Let us dive into missing data. Missing data occurs, when no value is stored in some of the columns.

Let us take an example: say someone is conscious about his/her weight and doesn't fill the weight field in a survey. Then, the weight value for that certain person will be missing.
someoneある人
conscious about意識する
certain特定の
Thenその結果

How Pandas Handles missing data

Pandas handles missing values in two ways. The first you've seen before in previous sections: NaN, or Not a Number. This is a actually a special value that is part of the IEEE floating-point specification and it is only used to indicate missing floating-point values.

For missing values apart from floats, pandas uses the Python None object. While it might seem confusing that you will encounter two different kinds of values that say essentially the same thing, there are sound programmatic reasons for this design choice and, in practice, going this route enables pandas to deliver a good compromise for the vast majority of cases. Notwithstanding this, both None and NaN carry restrictions that you need to be mindful of with regards to how they can be used.
before前から
apart from以外は
Notwithstanding thisそれにもかかわらず
restrictions制限
carry持っている



None: non-float missing data
Because None comes from Python, it cannot be used in NumPy and pandas arrays that are not of data type 'object'. Remember, NumPy arrays (and the data structures in pandas) can contain only one type of data. This is what gives them their tremendous power for large-scale data and computational work, but it also limits their flexibility. Such arrays have to upcast to the “lowest common denominator,” the data type that will encompass everything in the array. When None is in the array, it means you are working with Python objects.

To see this in action, consider the following example array (note the dtype for it):
object 型とは、ざっくり言うと「文字や雑多なデータを入れる箱」のこと
tremendousすごましい
NumPyは、「同じ種類のデータ」だけを並べるから爆速
なのに None みたいな別種類のデータを混ぜると、全員を無理やり「何でもありの遅い型」に合わせちゃう
だから、数字のデータの中に None を混ぜちゃダメだよ！


In [ ]:
import numpy as np
#np.array()
#NumPy配列になると、普通のリストではできなかった「全ての数字を一気にまとめて計算する」という超便利な技（ブロードキャスト）が使えるようになる
example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

The reality of upcast data types carries two side effects with it. First, operations will be carried out at the level of interpreted Python code rather than compiled NumPy code. Essentially, this means that any operations involving Series or DataFrames with None in them will be slower. While you would probably not notice this performance hit, for large datasets it might become an issue.
Series（1列のデータ）や DataFrame（表データ）などの中にnoneが含まれると計算スピードが格段に落ちる

The second side effect stems from the first. Because None essentially drags Series or DataFrames back into the world of vanilla Python, using NumPy/pandas aggregations like sum() or min() on arrays that contain a None value will generally produce an error:
drag back後ろに引きずり戻す
rather than~ ~ではない
Series（1列のデータ）
DataFrame（表データ）
involve　含む
erformance hit性能やスピードが落ちること
side effect副作用
stems from生じる

In [11]:
example1.sum()

TypeError: unsupported operand type(s) for +: 'int' and 'NoneType'

Key takeaway: Addition (and other operations) between integers and None values is undefined, which can limit what you can do with datasets that contain them.
integer整数

unsupported operand type(s) for +サポートされない、演算されるデータの種類、+にとって

NaN: missing float values
In contrast to None, NumPy (and therefore pandas) supports NaN for its fast, vectorized operations and ufuncs. The bad news is that any arithmetic performed on NaN always results in NaN. For example:
performed on実行される

In [12]:
np.nan + 1

nan

弱点、四則演算でnanがあると結果がnanになる

In [13]:
np.nan * 0

nan